test/integration/version_upgrade_test.go

```diff
@@ -35,6 +35,113 @@ import (
+// 2. Upgrade with newer version
+//   ./out/minikube start -p vupgrade --kubernetes-version=v1.13.0
+// 3. Downgrade with older version (this will spit out error and test the error message)
+//   ./out/minikube start -p vupgrade --kubernetes-version=v1.13.0
+func TestDifferentVersionUpgrade(t *testing.T) {
+	MaybeParallel(t)
+	WaitForStartSlot(t)
+	profile := UniqueProfileName("testkubernetestversions")
+
+	ctx, cancel := context.WithTimeout(context.Background(), 55*time.Minute)
+	defer CleanupWithLogs(t, profile, cancel)
+
+	// create a temporary file
+	tf, err := ioutil.TempFile("", "minikube-release.*.exe")
+	if err != nil {
+		t.Fatalf("tempfile: %v", err)
+	}
+	defer os.Remove(tf.Name())
+	tf.Close()
+
+	// download minikube
+	url := pkgutil.GetBinaryDownloadURL("latest", runtime.GOOS)
+
+	// chmod the binary (when Linux)
+	if runtime.GOOS != "windows" {
+		if err := os.Chmod(tf.Name(), 0700); err != nil {
+			t.Errorf("chmod: %v", err)
+		}
+	}
+
+	// download the file
+	if err := retry.Expo(func() error { return getter.GetFile(tf.Name(), url) }, 3*time.Second, 3*time.Minute); err != nil {
+		t.Fatalf("get failed: %v", err)
+	}
+
+	t.Run("parallel", func(t *testing.T) {
+		tests := []struct {
+			name                 string
+			message              string
+			kubernetesversion    string
+			expectstartupfailure bool
+		}{
+			{"fresh install kubernetes v1.13.0", "v1.13.0", "v1.13.0", false},
+			{"upgrade kubernetes to v 1.14.0", "v1.14.0", "v1.14.0", false},
+			{"downgrade kubernetes to v 1.12.0", "Non-destructive downgrades are not supported", "v1.12.0", true},
+		}
+		for _, tc := range tests {
+			tc := tc
+			t.Run(tc.name, func(t *testing.T) {
+				// instruct to install kubernetes v1.13.0
+				args := append([]string{"start", "-p", profile, fmt.Sprintf("--kubernetes-version=%s", tc.kubernetesversion), "--alsologtostderr", "-v=1"}, StartArgs()...)
+
+				rr := &RunResult{}
+				var startingFunc = func() error {
+					rr, err = Run(t, exec.CommandContext(ctx, tf.Name(), args...))
+					return err
+				}
+
+				// if the test expect failure the function is different
+				if tc.expectstartupfailure {
+					startingFunc = func() error {
+						rr, err = Run(t, exec.CommandContext(ctx, tf.Name(), args...))
+						stderr := rr.Stderr.String()
+						if !strings.Contains(stderr, tc.message) {
+							t.Fatalf("test failed as no message %s detected", tc.message)
+							return err
+						}
+						return nil
+					}
+				}
+
+				err := retry.Expo(startingFunc, 1*time.Second, 30*time.Minute, 3)
+
+				// as the text expect failure there is no need to continue the flow
+				if tc.expectstartupfailure {
+					return
+				}
+
+				if err != nil {
+					t.Fatalf("release start failed: %v", err)
+				}
+
+				// grab the Stdout for checking...
+				ss := rr.Stdout.String()
+
+				// fail if it does not contain the string
+				if !strings.Contains(ss, tc.message) {
+					t.Fatalf("test failed as no message %s detected", tc.message)
```

Comment:
The stdout of 'minikube start' is not designed to be parsed. Do you mind checking for the resulting apiserver version of the cluster that was started instead?

Something like:

```

	j, err := exec.Command(path, "version", "--client", "--output=json").Output()
	if err != nil {
		return errors.Wrap(err, "exec")
	}

	cv := struct {
		ServerVersion struct {
			GitVersion string `json:"gitVersion"`
		} `json:"serverVersion"`
	}{}
	err = json.Unmarshal(j, &cv)
	if err != nil {
		return errors.Wrap(err, "unmarshal")
	}
```

Something like what we do at:

https://github.com/kubernetes/minikube/blob/1c46502f358e791c7a7a1ce708c7ff63480f3697/cmd/minikube/cmd/start.go#L519